test/test_broker.hpp

```diff
@@ -1559,44 +1603,16 @@ class test_broker {
             // other than to enforce the uniqueness constraints.
             // Potentially this can be enabled only in debug builds.
             mi::ordered_unique<
-                mi::tag<tag_con_topic>,
+                mi::tag<tag_con_topic_filter>,
                 mi::composite_key<
                     sub_con_online,
                     BOOST_MULTI_INDEX_MEMBER(sub_con_online, con_sp_t, con),
-                    BOOST_MULTI_INDEX_MEMBER(sub_con_online, MQTT_NS::buffer, topic)
+                    BOOST_MULTI_INDEX_MEMBER(sub_con_online, MQTT_NS::buffer, topic_filter)
```

Comment:
You store now the complete topic filter to prevent duplicates. But the subscription_map could actually help you with that. But then the pointer should be to a session, and once insert a subscription again to the same session, in the set the session was already there. This way also you don't have to store the complete topic next to having this already in the subscription map.

Reply:
I understand. The current design separates session and subscription. I use session_map just for quick access.
I keep the current design, so far.

Reply:
I think about that. 
What happened the following case ?

1. session1 subscribe `a/+/c` QoS1
2. session1 subscribe `a/b/+` QoS2
3. publish `a/b/z` QoS2
   * The broker needs to deliver QoS2 message to session1.
     How do I know QoS2 ?
4. publish `a/z/c` QoS2
   * The broker needs to deliver QoS1 (min(QoS2, QoS1)) message to session1.
     How do I know QoS1 ?

```cpp
m.find(
    "a/b/z",
    [](session_state const& v) {
        // Step 3 and Step 4 have the same v. How to distinguish QoS2 or QoS1
        // calling v.non_const_member_function(); may be required. How to do that ?
    }
);

Reply:
yes, i strugled with that also when implementing my test broker I have send you. You have two options:

- Send the message twice: once with QoS1, once with QoS2
- Calculate the highest QoS for each session and publish the message with the highest possible QoS once

I think I did the last, when I was editing you broker test. First, collected all sessions that match the topic, and per session calculate the QoS, then finally send the message once to every connection. 

I do not know what the MQTT standard specifies on this

Reply:
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901120


When Clients make subscriptions with Topic Filters that include wildcards, it is possible for a Client’s subscriptions to overlap so that a published message might match multiple filters. In this case the Server MUST deliver the message to the Client respecting the maximum QoS of all the matching subscriptions [MQTT-3.3.4-2]. In addition, the Server MAY deliver further copies of the message, one for each additional matching subscription and respecting the subscription’s QoS in each case.

Reply:
> the maximum QoS of all the matching subscriptions

Thank you @jonesmz

Reply:
Technically, either of your suggested approaches are correct because of this sentence from the standard:

> In addition, the Server MAY deliver further copies of the message, one for each additional matching subscription and respecting the subscription’s QoS in each case.

Reply:
> Personally I think that the best approach is to deliver the message only once. It has the same cost of iterating all matching subscriptions, but sends substantially less data over the network.

I agree

Reply:
I create #701 to continue the discussion.
I think that it is easy to search and refer it later.